src/Simulation/AutoSubstitution/RewriteStep.cs

```diff
@@ -0,0 +1,170 @@
+
+            // collect all callables that have an substitution attribute
+            var globals = globalCallables.Where(p => p.Value.Source.CodeFile.EndsWith(".qs"))
+                                         .Where(p => p.Value.Attributes.Any(HasSubstitutionAttribute));
+
+            if (!globals.Any())
+            {
+                diagnostics.Add(new IRewriteStep.Diagnostic {
+                    Severity = DiagnosticSeverity.Info,
+                    Message = "AutoSubstitution: no operations have @SubstitutableOnTarget attribute",
+                    Stage = IRewriteStep.Stage.Transformation
+                });
+                return true;
+            }
+
+            // no need to generate any C# file, if there is no substitution attribute, or if we cannot retrieve the output path
+            if (!AssemblyConstants.TryGetValue(Microsoft.Quantum.QsCompiler.ReservedKeywords.AssemblyConstants.OutputPath, out var outputPath))
+            {
+                diagnostics.Add(new IRewriteStep.Diagnostic {
+                    Severity = DiagnosticSeverity.Error,
+                    Message = "AutoSubstitution: cannot determine output path for generated C# code",
+                    Stage = IRewriteStep.Stage.Transformation
+                });
+                return false;
+            }
+
+            diagnostics.Add(new IRewriteStep.Diagnostic {
+                Severity = DiagnosticSeverity.Info,
+                Message = $"AutoSubstitution: Generating file __AutoSubstitution__.g.cs in {outputPath}",
+                Stage = IRewriteStep.Stage.Transformation
+            });
+
+            using var writer = new StreamWriter(Path.Combine(outputPath, "__AutoSubstitution__.g.cs"));
+            var context = CodegenContext.Create(compilation, AssemblyConstants);
+
+            var generator = new CodeGenerator(context);
+            foreach (var (key, callable) in globals)
+            {
+                var attributeArguments = callable.Attributes.Where(HasSubstitutionAttribute).Select(GetSubstitutionAttributeArguments);
+                foreach (var (alternativeOperation, _) in attributeArguments)
+                {
+                    var period = alternativeOperation.LastIndexOf('.');
+                    if (period == -1)
+                    {
+                        diagnostics.Add(new IRewriteStep.Diagnostic {
+                            Severity = DiagnosticSeverity.Error,
+                            Message = $"AutoSubstitution: name of alternative operation in {key.Namespace}.{key.Name} must be completely specified (including namespace)",
+                            Stage = IRewriteStep.Stage.Transformation
+                        });
+                        return false;
+                    }
+
+                    var qualifiedName = new QsQualifiedName(alternativeOperation.Substring(0, period), alternativeOperation.Substring(period + 1));
```

Comment:
To clarify, does this restrict the substitution target and alternative operation to be in the same namespace? Can `alternativeOperation` contain `.` characters to indicate a subnamespace of that namespace?

Reply:
They can be in different namespaces. Here, I am just splitting the string into namespace + operation name and construct a `QsQualifiedName` instance from the two parts.